src/core/PICA/dynapica/shader_rec.cpp

```diff
@@ -0,0 +1,31 @@
+#include "PICA/dynapica/shader_rec.hpp"
+#include "cityhash.hpp"
+
+#ifdef PANDA3DS_SHADER_JIT_SUPPORTED
+void ShaderJIT::reset() {
+	cache.clear();
+}
+
+void ShaderJIT::prepare(PICAShader& shaderUnit) {
+	shaderUnit.pc = shaderUnit.entrypoint;
+	// We construct a shader hash from both the code and operand descriptor hashes
+	// This is so that if only one of them changes, we still properly recompile the shader
+	// This code is inspired from how Citra solves this problem
+	Hash hash = shaderUnit.getCodeHash() ^ shaderUnit.getOpdescHash();
```

Comment:
This is a total nit-pick but xor is a commutative operation so behavior like `a^b == b^a` could produce the same hash here causing a possible collision-hazard. It's ridiculously low of course but you might want to do something like `std::rotl(x, 1) ^ y` to be more correct here.

Reply:
Very good observation. I'll think about it for a sec and change it.

Reply:
Added